integr8ly/keycloak-operator
# create monitoring resources

## Motivation

The keycloak operator should also create the resources required to enable monitoring for the keycloak instance.

## What

This PR adds phases to the operator for installing the monitoring resources (dashboard, service monitor, alerts). This might fail if the CRDs for those resources don't exist yet. In that case the operator will print a warning and retry.

## Verification Steps

1. Use a cluster that does not have any monitoring CRDs installed
1. Deploy keycloak using the operator (and this branch)
1. Keycloak should deploy fine but you should see warnings in the operator logs: `WARN[0060] Cannot create monitoring resource 'grafana-dashboard' at this time, retrying`
1. Now create the monitoring CRDs:
4.1 Grafana dashboard: https://github.com/integr8ly/grafana-operator/blob/master/deploy/crds/GrafanaDashboard.yaml
4.2 Prometheus rule: https://github.com/integr8ly/installation/blob/master/evals/roles/middleware_monitoring/templates/prometheus_rule_crd.yml.j2
4.3. Service monitor: https://github.com/integr8ly/installation/blob/master/evals/roles/middleware_monitoring/templates/prometheus_service_monitor_crd.yml.j2
1. The operator will now print out logs that creating the monitoring resources was successful.
1. Verify that those resources have been created.

You can also test it with the full installer by:

1. Updating the repo and branch to point to this PR here: https://github.com/integr8ly/installation/blob/master/evals/roles/rhsso/defaults/main.yml#L27
1. Updating the image to `pb82/keycloak-operator:monitoring` here: https://github.com/integr8ly/installation/blob/master/evals/roles/rhsso/defaults/main.yml#L2

## Progress

- [ ] Verification
